README.md

```diff
@@ -345,24 +345,73 @@ explained in terms of the sequential model. This is useful for trying to find
+         of this as an initialisation bit that setups up some state);
+       - a second part (the _suffix_) that will be split in sublists which will
+         be run in parallel (see `parallelSafe` to understand how it determines
+         that a sequence of commands can be run in parallel). More than one
+         suffix can be generated, i.e. this second step can be done multiple
+         times with the part of the generated list that doesn't belong to the
+         prefix.
 
-  3. execute the prefix sequentially;
+  2. execute the prefix sequentially as in the section above (checking pre and
+     post-conditions);
 
-  4. execute the suffixes in parallel and gather the a trace (or history) of
-     invocations and responses of each action;
+  3. execute the suffixes in parallel without checking pre/post-conditions
+     and gather the trace (or history) of invocations and responses of each
+     action;
 
-  5. try to find a possible sequential interleaving of action invocations and
-     responses that respects the post-conditions.
+```
+                            ┌── no checks aside from ensuring no exception was thrown
+                            │
+                  ╭─────────┴──────────╮
+
+                  ┌─ [C] ──┐  ┌ [F, G] ┐   ◀─╮
+Commands: [A, B] ─┤        ├──┤        │     ├─ executed `concurrently`
+                  └ [D, E] ┘  └ [H, I] ┘   ◀─╯
+
+          ╰─┬──╯       ▲          ▲
+            │          ╰─────┬────╯
+            │                └── groups are not run in parallel
+            │                    i.e [C, D, E] will run (and
+            │                    finish) before F or H are
+            │                    started
+            │
+            └── pre/postconditions and invariant checked
+                executed sequentially
+```
+
+  4. if something goes wrong when executing the commands, shrink the generated
+     commands and present a minimal counterexample;
+
+  5. otherwise, try to find a possible sequential interleaving of action
+     invocations and responses that respects the post-conditions. For each
+     interleaving, this is done by advancing the `Concrete` model (starting at
+     `initalModel`) through the sequence of pairs of invocations to `command
+     Concrete` and the returned `response Concrete` emitted at step 3, check the
+     postcondition at each pair.
+
+  6. if no possible sequential interleaving was found, then shrink the generated
+     commands and present a minimal counterexample.
 
 The last step basically tries to find
 a [linearisation](https://en.wikipedia.org/wiki/Linearizability) of calls that
 could have happend on a single thread.
 
+Notice that step 5 above introduces a subtlety in the post-condition checks and
+transitions for the `model Concrete`. Despite the system under test running in a
+concurrent way, the model can still be designed to work sequentially as *it will
```

Comment:
Yes, this is a key point of linearisability checking. It's also an assumption: the model most be correct with respect to sequential execution before it's used in concurrent testing.

Reply:
This is an important point, in my last commit I added this to the paragraph.